### ISSUE TITLE
Intity Epubreader

### ISSUE THREAD

**Issue by dunxd** [mikespub-org/intity-epubjs-reader#2]
I really like the look of the inity epub reader. I've taken a chance and made it the epubreader for HA-COPS.

Really nice table of contents, and bookmarks and notes!  However, where does it store the bookmarks and notes?  Looks like Local Storage in the browser.  Very cool, but how persistant?

**Comment by mikespub** [comment]
Hi @dunxd glad you like it :-)

I used the [fork from intity](https://github.com/intity/epubjs-reader) as starting point because its distribution was kept up to date, but the original is still the `epub.js` reader from [futurepress](https://github.com/futurepress/epubjs-reader)

Right now the "integration" with COPS is limited to 1 template + 1 zipfs.php file, with the actual functionality provided by the `epubjs-reader` package, so any changes would need to come from the upstream repos. I haven't explored all its capabilities, but from what I can see local storage is what you use by default.

For annotations there's an option to share them with [hypothes.is](https://web.hypothes.is/). Calibre has its own annotations table, but I haven't seen any example of integrating them with epubjs-reader. And I don't know if anything is available for bookmarks.

**Comment by mikespub** [comment]
Maybe one more remark:

COPS has always taken a "read-only" approach for the Calibre database, and it has no custom database of its own. If we want to save/restore bookmarks and/or annotations with epubjs-reader (beyond local storage in the browser), it would need to be:
1. with a third-party service like hypothes.is
2. using the Calibre annotations table (with updates?)
3. using the EPUB file annotations (?)
4. using e-reader specific annotations from Kindle, Kobo etc.
5. using a COPS database to handle that
6. ...

From PHP design perspective, I'm not a fan of any of those myself, except maybe options 1 or 2 - what do you think?

**Comment by dunxd** [comment]
I switched it on yesterday. Unfortunately it doesn't render nicely in a
mobile browser. See attached. All pages look like that in all books. This
is in Chrome on an Android phone. I suspect it is a display port issue.
Changing the width setting in Intity doesn't seem to fix it.

My preference now would be to have an option in the customise page, and set
this to default to Intity - on a mobile browser if someone cared they could
switch back to Monocle. If you agree this is a way forward I'll work on a
pull request.

Or maybe there is a fix to Intity to handle mobile browsers.

**Comment by mikespub** [comment]
Hi @dunxd I'm afraid the screenshot didn't make it via email.

Some settings you can change in the epubjs-reader interface itself, like:
- Layout: Paginated vs Scrolled
- Page width: 800 or less/more
- ...
And other options could probably be configured if needed as well, like the examples at https://github.com/futurepress/epub.js/tree/master/examples

But I haven't tried most of those out, so if there's a particular configuration that works well for you let me know :-)

**Comment by dunxd** [comment]
![Screenshot_20240312_133612_Chrome](https://github.com/mikespub-org/seblucas-cops/assets/16608541/7aa6f53c-5a3d-4204-8b4a-0ea809f7c06c)

Here is the screenshot. Every page looks like this on a mobile phone. 

Changing the width in the Intity settings from 800 to 100 doesn't change it, nor does changing the font size to 50%. This is when viewing the book with Chrome on a Samsung Android 14 phone.

**Comment by mikespub** [comment]
Huh - that's definitely not what we're looking for here :-(

I'm not familiar with CSS but there seem to be some hard-coded calculations in main.css of the intity epubjs-reader like
https://github.com/intity/epubjs-reader/blob/e9e7b8adb4246b9893ce1c71d3e4972b5862a1f2/dist/css/main.css#L58

If you do find a work-around, it might be worth sending a PR to https://github.com/intity/epubjs-reader

Otherwise you might be better off with Monocle for mobile viewing...

**Comment by dunxd** [comment]
I'll find some time to look into this, and if I find a fix, I will
contribute to Intity and hopefully it works its way into COPS that way.
That's better for everyone right?

In the meantime I'm reverting HA-COPS to monocle until we have this fixed.

**Comment by marioscube** [comment]
I really like the new epub-reader over monocle.

Tomorrow I can test it with iOS ipadOS if you want.

What I would really like in this epubreader is the ability to set the font-size (to larger) if possible.

**Comment by mikespub** [comment]
That'd be great, thanks. You're supposed to be able to change the font size via the settings in the sidebar on the left

**Comment by marioscube** [comment]
Thank you! I found the font setting in the sidebar. Just what I needed.

Then on to my findings with iPhone and iPad.

1. I can confirm the issue of @dunxd on my iPhone 13 in portrait mode with all webkit browsers.
2. On the iPhone in Safari in landscape mode it works (sort of). The viewing area for the book is to narrow for the screen. It can be used for a quick look at a book, not really for reading. 
3. On an iPad it works (again sort of) in portrait and landscape mode. To be able to use the settings and then read again I had to juggle the iPad between landscape (I use this) and portrait mode. In landscape the setting options would sometimes disappear off the screen.
4. The viewing rectangle is not wide enough for comfortable reading on the iPhone in landscape mode. On the iPad it's OK, but could be slightly wider to optimally use the screen size. I have that same observation with a regular browser on my desktop.

**Comment by mikespub** [comment]
Thanks for the feedback - I guess my hope of finding a decent replacement for monocle isn't quite realized yet...

If anyone knows of other potential candidates please let me know.

Pre-conditions would be:
1. stand-alone package that requires minimal customization of html template and/or javascript to integrate
2. actively maintained or updated on a regular basis in the last x years
3. allows setting the initial book path in template/script and retrieve individual epub file components (for larger epub files)
4. visually attractive & functional for both large screens, mobile devices and e-readers
5. (bonus) possibility of loading/saving bookmarks, annotations etc.

Note: that last part seems to be a dead end so far, as apparently every viewer uses its own variation on the EPUB CFI standard - see comments and links at https://github.com/futurepress/epub.js/issues/1358

**Comment by marioscube** [comment]
@mikespub 

Take a look at ../cops/vendor/mikespub/epubjs-reader/dist/css/main.css
On the bottom of this file remove /* \*/ around the code below \/* For iPhone portrait layouts only \*/ and /\* For iPhone landscape layouts only */, remove the cookies from Safari on my iPhone and ..... it works! (on iPhone) :-)

Not 100% perfect, but now it is usable.

@dunxd 
See if this works on Android.

**Comment by dunxd** [comment]
Hmm - that's some interesting CSS going on.  Not only in the CSS files, but I think there are some styles that are written in via the CSS at least to generate the columns that are displayed when the viewport is over a certain width.

Lines 58 and 59 of `/cops/vendor/mikespub/epubjs-reader/dist/css/main.css` calculate how tall and wide the container for the book should be by subtracting some figures from the rendered width.  I think these are set rather high so I adjusted them as follows:
```css
	width: calc(100% - 50px);
	height: calc(100% - 90px);
```
I just made this change using the inspector in Chrome, and then resized the browser window to see the effect at different sizes. At first glance that seems to work ok for smaller screens. On larger screens it starts to render the page in multiple columns so it looks like a book, and then on a wide screen it starts to look a bit too wide, which I guess is why they made them so large.  I also don't know what it will look like when changing the font size and width setting in the reader's menu.

With much fiddling it might be possible to get this working better using @media directives in the CSS to have different sets of these values for different screen sizes.  I may find some time this weekend to look at this deeper, but if you try the above changes to main.css and let me know what you see that would help me know if I am onto the right thing.

**Comment by dunxd** [comment]
If I fork Intity's epubjs-reader how do I then get my fork of _this_ repo to use it so I can test changes I make.  I have still not figured out composer. 

I fear that I will eat up a lot of time just preparing to experiment - any pointers?

**Comment by mikespub** [comment]
Hi @dunxd

here's a short description: https://getcomposer.org/doc/05-repositories.md#loading-a-package-from-a-vcs-repository

What I tend to do for minor tweaks in libraries is
1. first change the files directly in the vendor/ directory of the main project until I'm satisfied,
2. then fork the library repo (*), create a new branch and apply/commit the same changes there.
3. update the composer.json file of the main project to use your forked branch of the library (as described above)
4. and try out the applied changes by running `composer update` again
5. submit your forked branch as PR to the library maintainer (*)

(*) you'll need to fork from <https://github.com/mikespub-org/intity-epubjs-reader> if you want to update the package, since <https://github.com/intity/epubjs-reader> doesn't have a composer package definition

In this case, once you've done step 5 I can merge it into my fork too, and have it required in future COPS versions

**Comment by marioscube** [comment]
@dunxd @mikespub 

I player around with some of the settings in ../cops/vendor/mikespub/epubjs-reader/dist/css/main.css and I like for webbrowser (firefox), iPad and iPhone:

change
#viewer : width: calc(100% - 416px);
#viewer : height: calc(100% - 116px);
.arrow span : font-size: 64px;
#divider : height: calc(100% - 116px);

to
#viewer : width: calc(100% - 100px);
#viewer : height: calc(100% - 100px);
.arrow span : font-size: 24px;
#divider : height: calc(100% - 100px);

The latest changes in https://github.com/intity/epubjs-reader make it look slightly nicer.

On a very wide screen (I have a 3840x1600 monitor) it's perhaps a bit to much to comfortably read a book. ;-)

**Comment by dunxd** [comment]
I've created a PR on https://github.com/mikespub-org/intity-epubjs-reader which changes the CSS to detect screens with the same or smaller width as an iPhone X, and makes the viewer wider (100% - 60px) and the page turn arrows smaller.  Thanks @marioscube for the suggestion - I've made it wider than you suggested.  I've tested on my Samsung Galaxy A53 and it works great.  Would be good to get feedback.

Here is the css:
```css
@media only screen and (max-width: 812px){ /* This should catch smart phones with a screen width of, or narrower than, an iPhone X */
    #viewer {
        width: calc(100% - 60px);
    }
    .arrow span {
        font-size: 24px;
    }
}
```
I didn't notice any difference for the height settings or the divider, so I took them out.

Would love to hear how it works with your own phones.  I'm not sure whether the same width is detected if the phone is in portrait or landscape mode.  The rule could be further modified to detect that, but I wanted to start simple and get it right for the most common use case first.

**Comment by dunxd** [comment]
I've done some more tweaking to my PR.  If you want to try it locally, it is now:
```css
@media only screen and (max-width: 812px){ /* This should catch smart phones with a screen width of, or narrower than, an iPhone X */
    #viewer {
        width: calc(100% - 60px);
        height: calc(100% - 60px);
        margin-top: 45px;
        box-shadow: 0 0 5px rgba(0,0,0,.12), 0 5px 5px rgba(0,0,0,.24);
    }
    .arrow span {
        font-size: 24px;
    }
}
```
This reduces the wasted space at the bottom of the screen, and looks better in landscape mode too. This only applies to smaller screens, so no change for iPads etc.

**Comment by marioscube** [comment]
@dunxd 

Your second "tweak" is definitely better.

It works well on both an iPhone 6s and iPhone 13 in portrait mode.
Sometimes the "hamburger" icon disappears if the book title is to long. But with some switching between landscape and portrait mode that can be overcome.

On the iPhone 13 a margin-top of 30px looks slightly better. as does a width pf -80px. But that's a personal preference.

In landscape mode this PR works as well on both the iPhone's.

The divider only needs to be adjusted is you change something with height and there are 2 columns.

**Comment by dunxd** [comment]
Thanks for feedback.  

I'm not seeing the hamburger icon disappearing for any screen width in Chrome browser or Safari on a Mac, so perhaps I don't quite understand what you mean.  The book title isn't displaying for me at all so I am not sure what would cause this to disappear. Could you share a screenshot?

I've put the divider back in so it handles the height setting for smaller screen.  You will only see this if you reduced the page width setting to narrow enough that it splits into columns on the phone screen - e.g. set it to 200.

100%-80px works does indeed look better with the visible page turn buttons.

I tweaked the top margin so the top menu appears more vertically centered, but that, as you say, is very subjective.

```css
@media only screen and (max-width: 812px){ /* This should catch smart phones with a screen width of, or narrower than, an iPhone X */
    #viewer {
        width: calc(100% - 65px);
        height: calc(100% - 60px);
        margin-top: 40px;
        box-shadow: 0 0 5px rgba(0,0,0,.12), 0 5px 5px rgba(0,0,0,.24);
    }
    .arrow span {
        font-size: 24px;
    }
    #divider {
        margin-top: 40px;
        height: calc(100% - 60px);
    }
}
```

Some ideas for further improvement:
* Drop the visible page buttons for invisible ones over the left and right of the text - you can tap anywhere in the 100 px right or left to turn pages. This allows more screen width for the words on the page.
* Rethink the top menu - the folder icon invites uploading a file to open - that isn't really necessary for COPS so the icon could be dropped.  The bookmark and fullscreen buttons are useful, but could they be incorporated into the hamburger menu instead, allowing more book display.
* Maybe it would be nice to display the book title and author at the top, with truncation for very long strings.

**Comment by mikespub** [comment]
Thanks for testing this out :-)

Regarding the book title, this was actually moved to a new metadata panel by @intity last week I believe - which might explain the confusion.

I've merged the pull request and added a new tag so that you can update the package via composer if you want...

**Comment by marioscube** [comment]
@dunxd 

See picture from my iPhone in portrait.
No "hamburger" menu icon. Title is to wide.

![image0](https://github.com/mikespub-org/seblucas-cops/assets/20281601/3da658f1-8239-4968-84f8-40b1e1b032da)

And at the bottom it does not look to nice.

**Comment by intity** [comment]
I think I need to clarify some details regarding the [epubjs-reader](https://github.com/intity/epubjs-reader) fork. This is actually an experimental project for me to learn some concepts of JavaScript application development. To use this reader on a mobile device, you need to thoroughly develop style sheets and implement the additional `sidebarreflow` function to automatically rebuild the page layout (see [sidebarreflow](https://github.com/intity/epubjs-reader/blob/master/src/reader.js#L122)). In addition, the epub.js library is poorly developed, which will require additional effort to maintain.
Thanks

**Comment by mikespub** [comment]
Thanks for the feedback @intity - it's good to know what you're aiming for :-)

As mentioned above https://github.com/mikespub-org/intity-epubjs-reader/issues/2 we are looking for a possible alternative for the very old Monocle-based epub reader that was included with COPS, and an `epub.js` based reader seemed like a good fit.

The reason I started epubjs-reader from your fork was because it was kept up to date and you seemed to actively work on it, as opposed to the original epubjs-reader repo from futurepress.

One of the challenges for COPS is that it's targeted not just for large screens, so if mobile devices or old e-readers are not on your roadmap at all, we may need to start from another fork or find some other starting point...

**Comment by marioscube** [comment]
As a user of COPS I must say that the new epub **viewer** based on epub.js is much, much better than monocle.

It works very well with my iPhone(s) and iPad(s). Maybe not for really reading an epub. I have epub-reader apps for that. But to just view an epub to see if I want to start reading it, it just works. It's a good addendum to COPS.

**Comment by dunxd** [comment]
I am with @marioscube on this - @intity has made great improvements on the futurepress epubreader, and I believe with some tweaks we can get it working well for our use case. However, we haven't been able to test it against many different devices so there could well be snags waiting to bite us.

What I get from this discussion is that the fork made by @intity is only being developed by them, and we should not hold any expectation about future development. That is fair enough.

Would it make sense to make the @mikespub fork from the @intity one instead a fork from the futurepress version, and incorporate intity's changes which we like (of course with full acknowledgement and thanks).?This would allow us to incorporate any improvements to futurepress version in future (if they happen) or even contribute PRs to futurepress. That stops the version we use being dependant on continued maintenance by intity.

Is that reasonable?

**Comment by intity** [comment]
@dunxd, the problem is that the developer from Futurepress does not maintain the main epub.js library well (for example, the metadata is not fully extracted).

I think the main question should be posed like this: who will maintain the epub.js library in accordance with the [EPUB 3](https://idpf.org/epub/30/spec/epub30-overview.html) specification?

**Comment by mikespub** [comment]
As I see it we have 2 separate steps here:

1. maintain/update **epub.js** itself, which gets imported into whichever epubjs-reader we use - is there a more recent fork from epub.js that you could require in the @intity repo package.json instead of `"epubjs": "^0.3.93"`?

2. maintain/update **epubjs-reader** to better suit the needs of COPS users - I'm already forking the @mikespub repo from the @intity repo to import as PHP package into COPS, but if minor tweaks in .css or .js are needed for COPS users, I'd prefer that they be merged into the @intity repo at some point, since I'm a total loss at either stylesheets or javascript ;-)

**Comment by dunxd** [comment]
Is adopting intity's epubreader.js without taking on the upstream epub.js reader any worse than sticking with monocle? It's not like monocle is more actively maintained. The last update was 7 years ago, and there are no forks with updates in the last five.

In my opinion, the Intity refer is a cool improvement on the current browser based reader. I also note that in my opinion the browser based reader isn't a core function of COPS. As @marioscube said it is a nice addendum. 

Would it be so bad if we just adopted @mikespub fork of Intity, tweaked it for mobile phones (I think we are 90% there already) and accepted there will be no upstream development of Intity or epub.js ever.

**Comment by intity** [comment]
@dunxd, I don’t know anything about the monocle reader, so I can’t give an objective assessment. You are right about browser technology. The main advantage is the ability to integrate epub reader into your website. In addition, the user can use the built-in Google translator to translate the book into other language cultures.
I experimented a little with the layout for a mobile device on an Android emulator and this is what I came up with (see [ebubjs-reader-to-pixel-7-pro](https://disk.yandex.ru/i/ijJatjUA8fYQMw)). All dimensions are set in pixels. I think that the sidebar should be placed on top of the "viewer" container, as is implemented in a mobile device. This way we will get a single layout for all devices. The main difference is that the "back" and "next" arrows will be located in the toolbar. This layout should be designed in such a way that the user can independently add buttons to the toolbar through the settings. In addition, if the buttons will not fit on the toolbar, they should be automatically hidden.

**Comment by intity** [comment]
@mikespub
1. Maintaining the epubjs library will take a lot of time. The most difficult task is to understand the architecture of the epubjs project. It also requires a deep understanding of the EPUB specification.
2. As far as I understand, you need epubreader to provide support for features for mobile devices. I'll try to implement this.

**Comment by mikespub** [comment]
Thanks @intity - I'd be very happy if you could help bring 2. along :-)

As for 1. I wasn't suggesting that you do it yourself, but simply check if some forked update of the last "official" epub.js version `"^0.3.93"` from @futurepress could be used in your project instead.

There seem to be a lot of forks available at https://github.com/futurepress/epub.js/forks anyway, but I don't know if any might be a better choice...

**Comment by intity** [comment]
I've added a new [theme](https://github.com/intity/epubjs-reader/tree/theme) branch with initial support for mobile devices.

Major changes

1. The side panel slides out from the left over the viewer container.
2. Navigation buttons "prev" and "next" are moved in the toolbar.
3. The user interface on the bookmarks and annotations panel has been unified.
4. Annotations are now added using a special input panel after highlighting range on the page.

Before we merge into the `master` branch, we need to test first.
Thanks

**Comment by mikespub** [comment]
Great, thanks @intity 

For those who want to test this @dunxd @marioscube I updated and re-packaged the @mikespub epubreader fork for composer

So you can apply
```shell
$ composer update -o
```
in your COPS directory to have the latest version (2024.3.31) working with COPS

**Comment by marioscube** [comment]
@mikespub 
$ composer update -o 
This did not work for me (installed cops from distribution)
$ php composer.phar update -o
Installed a shitload ;-) of directories in ../cops/vendor/mikesepub/
Do I need all of this?

The viewer feature does not work with Brave browser. With firefox it works.

@intity 
**Thank you for your update!**
It looks really good on my iPhone 13 and iPad 9. It does not work on an iPad 3 with iOS 8, but that is to be expected :-(
I will not even try it on my old Kobo Glo HD.

However, a big however, I hate the arrows at the top of the screen to page forward and back. This does not work on an iPhone. My thumbs simply can not reach it. Please revert to the arrows at the sides. It does not look nice, but it works!

In the browser desktop version it would be nice to be able to page forward and back with the keyboard arrow keys.

**Comment by intity** [comment]
@mikespub 
I also don't like the arrows on the top, but this solves the collision problem in the viewer container area. The ideal solution would be to turn pages using a `touchmove` event. I tested the `touchmove` event as an option, but the result looks terrible. Please clarify what exactly is not working on your iPhone? Can't click on the arrows or the page turn function isn't working?

**Comment by marioscube** [comment]
@intity 

1 - it looks great and modern,
2 - technically the arrows at the top work, but
3 - while reading I hold my iPhone (and iPad) by the sides or at the bottom. This way I cannot reach the top of the screen with my thumbs to press page forward or page back (the arrows)

Is it a strange idea to place the arrows (centered) and all the other icons on the bottom of the screen?

If you could make a touchmove work, that would be perfect.

**Comment by dunxd** [comment]
Monocle uses full height, set width empty DIVs at the left and right as turn buttons, allowing tapping on the right or left of the screen to turn the page forward and backwards.

Check out this Codepen to demonstrate what I mean and how to do it: https://codepen.io/dunxd/pen/GRLOEPr

This looks something like this, with the magenta areas representing the clickable parts of the page:
<img width="266" alt="image" src="https://github.com/mikespub-org/seblucas-cops/assets/16608541/cdec8668-9903-49b1-8849-ac9f4f21fe11">


I've set the cursor to change to left-right resize icon when hovering over either side of the page.

```css
#turn_back {
  position: absolute;
  display: block;
  left: 0px;
  width: 10%;
  height: 100%;
  z-index: 100;
}
#turn_forward {
  position: absolute;
  display: block;
  right: 0px;
  width: 10%;
  height: 100%;
  z-index: 100;
}

/* This block creates makes the entire div containing the link
 into a hyperlink, and changes the cursor to something close to a page turn indicator */
#turn_back a::after, #turn_forward a::after {
  content: "";
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  cursor: ew-resize;
 /* This is just to make the clickable area visible for demonstration purposes
 - it would be removed in the final version */
 background-color: rgb(255, 0, 255, 0.5);
}
```
Might something like this be preferable to visible links?

**Comment by intity** [comment]
@dunxd 
using overlay of hidden `prev` and `next` controls is a bad idea as it will create the problem of selecting text for adding annotations.

Another option is to change `manager` to `continuous` (see [continuous-scrolled](https://github.com/intity/epub.js/blob/master/examples/continuous-scrolled.html#L47)). This will make continuous scrolling in "scrolled" mode (without using arrows). However, I wouldn't want to enable this feature by default.

**Comment by marioscube** [comment]
@intity 

The question is "do you want epubjs-reader to be a **reader** or a **viewer**?"

It's your project, so you decide.

I personally would like to have a viewer to be able to quick scan a book/epub before i download it to my real epub-reader app or device (kobo). 

I do not do annotations, but some others can't live without it. 

I would like to keep COPS as minimalistic and utilitarian as possible with a good epub-viewer. 

Just my contribution to this discussion.

**Comment by horus68** [comment]
I'm also in favour of the option "epub-viewer" only!

**Comment by dunxd** [comment]
If the width of the touchable area is less than or equal to the padding of the container, then it should be possible to use this approach without preventing the selection of text for annotation. I've updated [the codepen](https://codepen.io/dunxd/pen/GRLOEPr) to show this.

The use case of the epub reader in COPS today seems to be more as a way to quickly see epub content, rather than a for lengthy reading sessions, but the annotation and bookmarking is _very_ cool and might be what some people need. It works fantastically well and since it keeps place and has bookmarks, it could make COPS more of a one stop shop. I wouldn't want to discard this feature.  Who knows where it might lead.

**Comment by marioscube** [comment]
@dunxd 

There is a difference in what I would like to have now and what might be in the future.

I'll go with what the developers want an let's encourage them (@mikespub and @intity) to keep improving COPS and epubjs-reader!

**Comment by intity** [comment]
It turns out that the epubjs library already has support for the page swipe feature. I set up the configuration this way: if the useragent is a mobile device, then the manager definition is `continuous`. Otherwise, the manager will be defined as `default`. Please test on your devices.

**Comment by intity** [comment]
Major changes for the [theme](https://github.com/intity/epubjs-reader/tree/theme) branch
5. By default, auto-hide arrows if the useragent is a mobile device.
6. Added `settings.controls` for reader configuration. For example,
```html
<script src="js/epubreader.js"></script>
<script type="text/javascript">
	const path = "https://s3.amazonaws.com/epubjs/books/alice.epub"
	window.onload = (e) => {
		epubreader.main(path, {
			restore: true,
			controls: {
				arrows: true,
				openbook: false,
				bookmarks: true,
				annotations: false,
				fullscreen: false
			}
		});
		e.stopPropagation();
	}
</script>
```
Please test on your devices.

**Comment by mikespub** [comment]
I updated my epubjs-reader fork with the new theme, adapted the template there and prepared a new release 2024.4.4 for composer.

In COPS I also had to adapt the [templates/epubjs-reader.html](https://github.com/mikespub-org/seblucas-cops/blob/main/templates/epubjs-reader.html) template file: https://github.com/mikespub-org/seblucas-cops/commit/34e1622952d0a2ec51d0907b2023e2543edeefae

You'll need to adapt the template (or update COPS) + run the `composer update -o` again to get the new features @intity prepared

Thanks for testing

**Comment by mikespub** [comment]
Included in new [COPS release 2.5.3](https://github.com/mikespub-org/seblucas-cops/releases/tag/2.5.3)

**Comment by marioscube** [comment]
@mikespub @intity 

I just did some testing with Firefox browser, iPhone 6s, iPhone 13, iPad 6th gen.

It all feels very "alpha". That is to say, it still needs "some" work ;-) But we're getting there!

1 - Browser
- paginated mode: works!
- scrolled mode: not tested.
Remark: Brave browser (chromium) gives strange results.

2 - iPad
- starts in paginated mode
- only way to page forward and back is through the arros at the top of the screen = unworkable
- paginated mode: works. Next and previous chapter trough arrows, which is ok. 
- paginated mode: screen is to wide in landscape mode (the way I usually use it). Protrait is better.

3 - iPhone's
- starts in paginated mode
- iPhone 6s: has not enough CPU power to work (very weak battery but even with a power bank attached it does not work)
- iPhone 13: want to work, but does not really. Page forward is usually more than one. Page back is even worse. Tapping does not work, only sliding.
- scrolled mode: works on iPhone 13 (not tested on iPhone 6s

General: the "open book icon" (map) does not work if book loaded through COPS. could be removed.

I still like the old option to tap to page forward/back in the middle of the page. Works best for now. For the iPhone scrolled mode is sort of ok as well.

Looking forward to the next "beta".

**Comment by mikespub** [comment]
Thanks for all the device testing @marioscube :-)

I'll leave the fixing to @intity @dunxd or whoever else can update the epubjs-reader, and stick to getting it to talk with COPS :-)

**Comment by mikespub** [comment]
Issue transferred to https://github.com/mikespub-org/intity-epubjs-reader to keep track of it

For other issues or comments on epubjs-reader (other than COPS integration), please check the original fork at https://github.com/intity/epubjs-reader

**Comment by marioscube** [comment]
@mikespub 

Good idea to make this a separate repository.

My only COPS related issue is that if epubjs-reader started from COPS can not (should not) open a new ebook through the map icon at the top of the screen.
This icon can be removed (for COPS).

**Comment by dunxd** [comment]
In my opinion moving the page turn controls to the top of the page is horrible.  It worked far better in the previous version with the arrows on the left and right of the content.  We can improve on that to get round overlaps by adjusting margin and padding widths. And it works the same way that most ereaders work, so its what people expect.

I'd love to contribute to making improvements, but I'm spending too much time trying to figure out how to juggle the latest version of COPS with an older version of the intity reader, leaving no time for doing the development work.

Changing the composer.json from 
```json
"mikespub/epubjs-reader": "^2024.4",
```
to 
```json
"mikespub/epubjs-reader": "^2024.3",
```
then running `composer update` isn't resulting in going back to where I consider things were in a better shape.

**Comment by intity** [comment]
@dunxd 
I agree with you, navigation buttons in the toolbar may not be the best idea. I'll put the arrows back. This will only work where the navigation arrows can fit within the margins of the indentation. For mobile devices, arrows will not appear by default.

Regarding the method of detecting mobile devices. To detect a mobile device, I use the [detectMobile](https://github.com/intity/epubjs-reader/blob/theme/src/reader.js#L180) function, which parses the `userAgent` string and applies a pattern of matches. There is also a more modern interface, [userAgentData](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/userAgentData), but this interface is currently experimental.

**Comment by dunxd** [comment]
I would not recommend using JavaScript to try and detect a mobile client. It won't remain accurate for long and user agent strings are not know for consistency or honesty, but the main reason is that it being a mobile client is less important than the size of the screen - and what works for a small screen also works for a small windowed browser on any platform. The Media Queries feature of CSS is the way to go in my opinion.

**Comment by intity** [comment]
@dunxd 
Mobile device detection is needed to definition two parameters:
1. [manager](https://github.com/intity/epubjs-reader/blob/theme/src/reader.js#L58)
2. display [arrows](https://github.com/intity/epubjs-reader/blob/theme/src/reader.js#L247)

**Comment by dunxd** [comment]
Does manager do anything other than force `continuous` display? That should be up to the user, not based on what class of device they are using.

Similarly, I'm not sure the desirability of visible arrows is any different on a mobile phone than anywhere else. They didn't look bad the way you had them previously, and I think they can be made responsive through simpler mechanisms than JavaScript browser detection.

I'm happy to do some work on this to try out different options - just run out of time this weekend.

**Comment by mikespub** [comment]
> 
> Changing the composer.json from
> ```json
> "mikespub/epubjs-reader": "^2024.4",
> ```
> to
> ```json
> "mikespub/epubjs-reader": "^2024.3",
> ```
> then running `composer update` isn't resulting in going back to where I consider things were in a better shape.

FYI, if you want to go back to one particular version, you can specify it directly in composer.jon (without the caret)
```json
"mikespub/epubjs-reader": "2024.3.31",
```

Or if you want to remain in the 2024.3.* family, you can specify one more level down `"^2024.3.5"` so it will upgrade/downgrade to the latest 2024.3.* but not go to 2024.4.* - see https://getcomposer.org/doc/articles/versions.md#caret-version-range-

**Comment by intity** [comment]
Major changes in the [theme](https://github.com/intity/epubjs-reader/tree/theme) branch:
1. Changes in configuration settings (see [#L196](https://github.com/intity/epubjs-reader/blob/theme/src/reader.js#L196)).
1.2. Properties for internal use are not saved in local storage (see [#L278](https://github.com/intity/epubjs-reader/blob/theme/src/reader.js#L278)). These properties are mainly used to pre-configure the reader during initialization and cannot be changed at the user interface level.
1.3. The properties `settings.annotations` and `settings.bookmarks` can take two values: `array` (default) or `false`.
1.4. There are three modes available for `settings.arrows`:
     - the `none` mode completely disables the arrows;
     - the `content` mode (default) adds navigation arrows in the middle of the page on the left `prev` and on the right `next`;
     - the `toolbar` mode adds arrows to the toolbar.

Example 1 (pre-config `flow` setting):
```html
<script type="text/javascript">
	const path = "https://s3.amazonaws.com/epubjs/books/alice.epub"
	window.onload = (e) => {
		epubreader.main(path, {
			flow: "scrolled",
			spread: {
				mod: "none",
				min: 800
			}
		})
		e.stopPropagation()
	}
</script>
```
Note that if the `flow` setting is set to `scrolled` you must also change `spread.mod` to `none`, otherwise it will not work. You also need to understand that these settings are restored from local storage when the page is reloaded.

Example 2 (pre-config `arrows` setting):
```html
<script type="text/javascript">
	const path = "https://s3.amazonaws.com/epubjs/books/alice.epub"
	window.onload = (e) => {
		epubreader.main(path, {
			arrows: "toolbar"
		})
		e.stopPropagation()
	}
</script>
```
This will force the toolbar arrows to be enabled for all devices, regardless of `userAgent` settings.

Example 3 (pre-config `annotations` and `bookmarks` settings):
```html
<script type="text/javascript">
	const path = "https://s3.amazonaws.com/epubjs/books/alice.epub"
	window.onload = (e) => {
		epubreader.main(path, {
			bookmarks: false,
			annotations: false
		})
		e.stopPropagation()
	}
</script>
```
This will completely disable the `annotations` and `bookmarks` UI.

Please test on your devices.

**Comment by mikespub** [comment]
New COPS [release 2.5.4](https://github.com/mikespub-org/seblucas-cops/releases/tag/2.5.4) using new intity-epubjs-reader [tag 2024.4.9](https://github.com/mikespub-org/intity-epubjs-reader/releases/tag/2024.4.9)

I added the epubjs-reader settings as template variable that can be configured via

$config['cops_epubjs_reader_settings'] = '{
    // ...
}';

in your COPS config_local.php file - see config_default.php for details.

That way you can experiment with the different settings via your config_local.php file - but some of them may be cached by the browser

**Comment by intity** [comment]
@dunxd 
> Does manager do anything other than force `continuous` display?

The epubjs library provides only two managers:
1. The `default` manager is used for desktop devices;
2. The `continuous` manager is used for mobile devices.

For manager (2) to work correctly in `paginated` mode, you also need to set the `snap` property to `true`. I can add a manager to the reader settings. Then you can apply your own mobile device detection feature.

**Comment by intity** [comment]
In order for me to merge with the `master` branch, I need to integrate the `Storage` class as a reader setting. The `Storage` class provides functions for storing a file from the device's file system in the browser's `window.indexedDB`. Therefore, when overriding the `window.onload` event, you need to set `settings.openbook` to `false`, otherwise you will get an error.

**Comment by mikespub** [comment]
Hi @intity I'm not sure I get your point here.

At this stage in COPS, I'm using template variables to fill in a template that looks like this (not updated here in theme yet):

```js
	<!-- Initialize bookPath and settings -->
	<script>
		// string containing the initial book path - link / to get content from file /{component}
		const bookPath = "{{=it.link}}";
		// js object containing the initial settings (use text for template variable)
		const settings = {{=it.settings}};
		window.onload = (e) => {
			epubreader.main(bookPath, settings);
			e.stopPropagation();
		};
	</script>
```

Where {{=it.settings} is replaced by the [text string defining the] js array in config_local.php:
```php
$config['cops_epubjs_reader_settings'] = '{
   // ...
}';
```

By default this will be an empty array, so if on your end you need to change the default handling of settings in epubjs-reader to cover some other issue, that's fine by me :-)

**Comment by mikespub** [comment]
BTW I'm thinking of loading the settings directly from a separate .json file instead in javascript, but that would mean another place where configurations are made.

**Comment by intity** [comment]
@mikespub 
I completely moved the `settings` to `index.html`. Essentially, `settings` is a javascript object that overrides the default reader settings. But it is important to understand the specifics of interaction with data that is restored from local storage when the reader is reloaded. That is, what is in the browser’s local storage has higher priority over pre-settings.

If you want to save session data in the server database, then this will require an additional API implementation to upload the settings object in json format to the server. Ideally, the server should provide the `application/json` API, and `epubreader.js` should interact with it as a frontend.

**Comment by mikespub** [comment]
Thanks @intity

in the COPS use case there's no problem if the browser local storage overrides whichever settings are set initially in the template.

After all, the idea is that COPS serves the books and epubjs-reader html/js/css files to the browser, and then epubjs-reader does its own thing in the browser :-)

So if after the initial page load the user wants to change settings in the browser etc. when reading a particular book, and have these settings re-used the next time he/she continues reading it, that's perfectly OK.

And COPS doesn't want or need to know what settings the user changed, so there's no need for a .json API to change/save settings - I was merely thinking to simplify the initial loading of the settings in index.html

**Comment by intity** [comment]
@mikespub 
Are you using the `openbook` feature for the COPS site?

I renamed the epubjs-reader repository to [epubreader-js](https://github.com/intity/epubreader-js). Please update the links.

**Comment by mikespub** [comment]
No, the open book feature is not relevant in this case, as @marioscube mentioned earlier. I'll have a look at the link later this weekend when I re-package a new version, thanks

**Comment by intity** [comment]
@marioscube
>I just did some testing with Firefox browser, iPhone 6s, iPhone 13, iPad 6th gen.

Please, if you are doing a test, you need to write more detailed information:
- Device name
- Operating system version
- Browser version

Ideally this should be the `userAgent` string. This data can be obtained by entering `navigator.userAgent` in the browser console. For an iPad/iPhone mobile device, `userAgent` data can also be obtained through the enabled **Web Inspector** option (see [Inspecting iOS and iPadOS](https://developer.apple.com/documentation/safari-developer-tools/inspecting-ios)).

**Comment by mikespub** [comment]
That would probably be @marioscube  rather than me :-)

**Comment by marioscube** [comment]
@intity @mikespub 

I will redo my "testing" probably over the weekend with the asked for information added.

To test I will use the latest version of COPS. (not the git clone version)

For devices I have:
 - iPhone 6s (with safari)
 - iPhone 13 (with safari)
 - iPad 6th generation (with safari)
 - iPad pro 1st generation (with safari)
 - macOS with firefox browser.
No android devices.

Any things I should especially test?

**Comment by intity** [comment]
@marioscube 
Can you get the `navigator.userAgent` strings for each device?

**Comment by marioscube** [comment]
@intity 

I am able to give you something like this from my apache logs:

Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/116.0.0.0 Safari/537.36

Is that what you want?

**Comment by intity** [comment]
@marioscube yes this is what I need.

**Comment by marioscube** [comment]
@intity 

OK that is easy, I just look at my (apache2) logs.

I could not find "navigator.userAgent" in the firefox console. Maybe I did something wrong. But now that's not needed.

**Comment by intity** [comment]
I did bundle optimization to reduce loading time. Also, I changed the package type to `module` (see [index.html](https://github.com/intity/epubreader-js/blob/master/dist/index.html#L22-L27) for current changes).
- More about [JavaScript modules](https://caniuse.com/?search=javascript%20modules)

Please test on your devices.

**Comment by marioscube** [comment]
I promised some quick testing and here it is:

### Server = Debian 12 Bookworm 
Webserver = Apache/2.4.59 (Debian)
PHP = PHP Version 8.2.18
COPS = COPS 2.5.4 - 20240409
	standard config_local.php


### Device = nucintosh 
OS = macOS 10.14.6
Browser = Firefox browser 115.10 (private browsing)
navigator.userAgent = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10.14; rv:109.0) Gecko/20100101 Firefox/115.0"

Font size (%) = 150
Layout = Paginated
Review = just works!

Layout = Scrolled
Review = Scrolled per chapter, then NEXT (key, mouse, menu) for next chapter


### Device = Apple iPhone 13
OS = iOS 17.4.1
Browser = Safari
navigator.userAgent = "Mozilla/5.0 (iPhone; CPU iPhone OS 17_4_1 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.4.1 Mobile/15E148 Safari/604.1"

Font size (%) = 150

Orientation = Portrait
Layout = Paginated (starts with)

Review = does not wrork properly for me
 - no arrows on top
 - no arrows to the side
 - browsing through swiping: 
	swipe left = nothing or 2 pages at once
	swipe right is nothing or back to start or 2 pages at once
=> this does not work

Orientation = Portrait
Layout = Scrolled
Review = Scrolles continuously through pages and chapters
	Works!

Orientation = Landscape
Layout = Paginated
Review = Does not work. No way to navigate the ebook. 
 - no arrows on top or to the side
 - swiping does not work for me
 - cannot reach all menu options

Orientation = Landscape
Layout = Scrolled
Review = Does not work properly
 - jumps forward and back with multiple pages while scrolling (swping up/down)
 - top bar of the reader takes to much space on the screen


### Device = Apple iPhone 6s
OS = iOS 15.8.2
Browser = Safari
navigator.userAgent = "Mozilla/5.0 (iPhone; CPU iPhone OS 15_8_2 like Mac OS X) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/15.6.6 Mobile/15E148 Safari/604.1"

Font size (%) = 150

Orientation = Portrait
Layout = Paginated (starts with)

Review = does not wrork properly for me
 - no arrows on top
 - no arrows to the side
 - browsing through swiping: 
	swipe left = nothing or 2 or more pages at once
	swipe right is nothing (sticks on same page) or back to start or 2 pages at once
=> this does not work

Orientation = Portrait
Layout = Scrolled
Review = Scrolles continuously through pages and chapters
	Works! (Some hickups caused by slow device?) 

Orientation = Landscape
Layout = Paginated
Review = Does not work. 
 - no arrows on top or to the side
 - swiping left works sort of
 - swiping left = stays on same page
 - cannot reach all menu options

Orientation = Landscape
Layout = Scrolled
Review = Does not work properly
 - sometimes jumps forward and/or back with multiple pages while scrolling (swping up/down)
 - top bar of the reader takes to much space on the screen


### Device = Apple iPad 9.7"
OS = iPadOS 17.4.1
Browser = Safari
navigator.userAgent = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/17.4.1 Safari/605.1.15"

Font size (%) = 150

Orientation = Portrait
Layout = Paginated 1 column (starts with)

Review = just works
 - no arrows on top
 - arrows to the side (love it)
 - icon for full screen that works
 - no browsing through swiping: 

Orientation = Portrait
Layout = Scrolled 1 column
Review = works
 - arrows to the side go to previous/next chapter
 - scrolling through swiping top/down works inside chapter 
 - icon for full screen that works

Orientation = Landscape
Layout = Paginated 
Review = works
 - no arrows on top
 - arrows to the side
 - swiping does not work
 - icon for full screen that works

Orientation = Landscape
Layout = Scrolled
Review = works 
 - arrows to the side go to previous/next chapter
 - scrolling through swiping top/down works inside chapter
 - icon for full screen that works


### Device = Apple iPad Pro 1st generation 12"
OS = iPadOS 16.7.7
Browser = Safari
navigator.userAgent = "Mozilla/5.0 (Macintosh; Intel Mac OS X 10_15_7) AppleWebKit/605.1.15 (KHTML, like Gecko) Version/16.6 Safari/605.1.15"

Font size (%) = 150

Orientation = Portrait
Layout = Paginated 2 column (starts with)

Review = just works (I personally do not like this mode)
 - no arrows on top
 - arrows to the side (love it)
 - icon for full screen that works
 - no browsing through swiping

Orientation = Portrait
Layout = Scrolled 1 column
Review = works
 - arrows to the side go to previous/next chapter
 - scrolling through swiping top/down works inside chapter 
 - icon for full screen that works
 - if I return to paginated mode, then 1 column!

Orientation = Landscape 
Layout = Paginated 2 columns
Review = works (best mode)
 - no arrows on top
 - arrows to the side
 - swiping does not work
 - icon for full screen that works

Orientation = Landscape
Layout = Scrolled 1 column
Review = works 
 - arrows to the side go to previous/next chapter
 - scrolling through swiping top/down works inside chapter
 - icon for full screen that works

**Comment by intity** [comment]
@marioscube 
Thank you for the detailed report. As I expected, your iPad device is detected as a desktop system in `userAgent`.

As you noticed, `manager=continuous` with `flow=paginated` does not work very well for a mobile device. This problem is also observed for desktop systems in the `manager=default` mode (for some books). This is a problem with the epubjs library.

**Comment by marioscube** [comment]
@intity 

Ah, that was what you were looking for!

Very deep in the Safari settings there is a setting REQUEST DESKTOP WEBSITE ON to which my settings on the iPad's say ON (slider to the right).
On my iPhone's the setting is OFF. 
If I set it to ON, I get the desktop website of COPS and the epubreader with arrow to the sides. If these were (a lot) smaller and the usable space on the screen for content was (a lot) larger I would be really happy. Maybe a setting?

**Comment by intity** [comment]
I think to get maximum performance from the **epubreder-js** application, you need to move towards component-based assembly. That is, the configuration should be configured not at the application level, but at the **webpack** or **rollup** builder configuration level. This way, everything unnecessary will be removed from the final application build. For example, if you disable the `openbook` feature, then the build system should remove all unnecessary code that the code analyzer spends time on.

**Comment by intity** [comment]
@marioscube 
This would require forcing arrows to be enabled for all devices, but I don't think that's the right solution. I can resize the arrows to adapt to a mobile device. However, changing the default settings requires careful strategy.

**Comment by dunxd** [comment]
I don't agree with using the User Agent string to figure out _anything_ about the device or the browser being used. It's terribly inconsistent, reliant on text parsing, can change with any update and prone to being spoofed for various reasons. [Mozilla strongly advise against it](https://developer.mozilla.org/en-US/docs/Web/HTTP/Browser_detection_using_the_user_agent).

Better to use other information that is more reliable such as display size (which also works for windowed browsers) and user configurable preferences. 

Just because I am reading on a phone doesn't mean I like continuous scrolling as opposed to turning pages. 

How are you going to handle ereader browsers should they ever get up to date CSS and JavaScript support? We already found that the User Agent strings for these is not very useful (e.g. no difference across completely different models produced 10+ years apart).

My suggestion is to use display size (easily accessed through CSS media queries) to apply settings likely to work well but override them with user preferences if they are set, and have reasonable fallbacks when neither apply.

**Comment by intity** [comment]
@dunxd 
It is not enough to get just the screen size, you also need to check whether the device supports touch features. This must be done before initializing the reader `settings`, since the [manager](https://github.com/intity/epubreader-js/blob/master/src/reader.js#L73) cannot be switched without reloading the page.

I also added manager to the settings so that the user can use his own function to detect the device. For example:
```html
<script type="module">
    import { Reader } from "./js/epubreader.min.js"
    const isMobile = () => { /* your code for mobile device detection */ }
    const url = "{{bookPath}}"
    window.onload = (e) => new Reader(url, {
        arrows: isMobile() ? "none" : "content",
        manager: isMobile() ? "continuous" : "default"
    })
</script>
```

**Comment by mikespub** [comment]
I've created a new release [2024.4.23](https://packagist.org/packages/mikespub/epubjs-reader#2024.4.23) of this package for composer + updated the COPS template accordingly: https://github.com/mikespub-org/seblucas-cops/compare/2.5.4...2.5.5

Any further tweaks of the arrows/manager/... settings for COPS can be changed in config_local.php if needed.

Thanks to everybody  for all the improvements and testing - I'm going to mark this issue closed for now.

Please open any new issues here or better yet, directly at https://github.com/intity/epubreader-js/issues :-)

**Comment by dunxd** [comment]
The version of Inity epubjs included with the current 2.6.1 release of COPS displays a blank screen when opening books in the browser on my android mobile. Was handling things based on the actual display port size ever implemented?
